core/src/test/scala/org/typelevel/paiges/Generators.scala

```diff
@@ -93,4 +94,28 @@ object Generators {
 
   implicit val cogenDoc: Cogen[Doc] =
     Cogen[Int].contramap((d: Doc) => d.hashCode)
+
+  implicit val shrinkDoc: Shrink[Doc] = {
+    import Doc._
+    def interleave[A](xs: Stream[A], ys: Stream[A]): Stream[A] =
+      if (xs.isEmpty) ys
+      else if (ys.isEmpty) xs
+      else xs.head #:: ys.head #:: interleave(xs.tail, ys.tail)
+    def combine[A](a: A, f: A => A)(implicit F: Shrink[A]): Stream[A] = {
+      val sa = shrink(a)
+      a #:: interleave(sa, sa.map(f))
+    }
+    def combine2[A](a: A, b: A, f: (A, A) => A)(implicit F: Shrink[A]): Stream[A] = {
+      val (sa, sb) = (shrink(a), shrink(b))
+      a #:: b #:: interleave(interleave(sa, sb), sa.flatMap(x => sb.map(y => f(x, y))))
+    }
+    Shrink {
+      case Union(a, b) => combine2(a, b, Union)
+      case Concat(a, b) => combine2[Doc](a, b, _ + _)
```

Comment:
Why does this need an ascription?  Can't scalac figure out `Doc` from the type of `a`?

Reply:
The anonymous function can't infer its type in this position.

If you declare `combine2` as something like:

```scala
def combine2[A](x: A, y: A)(f: (A, A) => A)(implicit F: Shrink[A]): Stream[A] = ...
```

...then you can probably avoid having to specify `Doc`. The reason for this is because it will set `A=Doc` for `x` and `y`, and then use that information to correctly infer the type of `f` without ascriptions.

Reply:
Interesting, thank you.  I'll do that.